In [ ]:
import pandas as pd



In [ ]:
df=pd.read_csv('/content/drive/MyDrive/DS/emails.csv')

In [ ]:
df.head()

text  spam
0  Subject: naturally irresistible your corporate...     1
1  Subject: the stock trading gunslinger  fanny i...     1
2  Subject: unbelievable new homes made easy  im ...     1
3  Subject: 4 color printing special  request add...     1
4  Subject: do not have money , get software cds ...     1

In [ ]:
df[df["spam"]==0]

text  spam
1368  Subject: hello guys ,  i ' m " bugging you " f...     0
1369  Subject: sacramento weather station  fyi  - - ...     0
1370  Subject: from the enron india newsdesk - jan 1...     0
1371  Subject: re : powerisk 2001 - your invitation ...     0
1372  Subject: re : resco database and customer capt...     0
...                                                 ...   ...
5723  Subject: re : research and development charges...     0
5724  Subject: re : receipts from visit  jim ,  than...     0
5725  Subject: re : enron case study update  wow ! a...     0
5726  Subject: re : interest  david ,  please , call...     0
5727  Subject: news : aurora 5 . 2 update  aurora ve...     0

[4360 rows x 2 columns]

In [ ]:
df.shape

(5728, 2)

**How many emails are in the dataset? - 5728**

In [ ]:
df[df["spam"]==1]

text  spam
0     Subject: naturally irresistible your corporate...     1
1     Subject: the stock trading gunslinger  fanny i...     1
2     Subject: unbelievable new homes made easy  im ...     1
3     Subject: 4 color printing special  request add...     1
4     Subject: do not have money , get software cds ...     1
...                                                 ...   ...
1363  Subject: are you ready to get it ?  hello !  v...     1
1364  Subject: would you like a $ 250 gas card ?  do...     1
1365  Subject: immediate reply needed  dear sir ,  i...     1
1366  Subject: wanna see me get fisted ?  fist  bang...     1
1367  Subject: hot stock info : drgv announces anoth...     1

[1368 rows x 2 columns]

**How many of the emails are spam? - 1368**


In [ ]:
word=[]
for i in df["text"]:
  word.append(i.split()[0].lower()[:-1])
print(set(word))  

{'subject'}


**Which word appears at the beginning of every email in the dataset? Respond as a lower-case word with punctuation removed.- subject**

**Could a spam classifier potentially benefit from including the frequency of the word that appears in every email?**

In [ ]:
a=0
for i in df["text"]:
  a=max(len(i),a)
print(a)  

43952


In [ ]:
sw = ("i", "me", "my", "myself", "we", "our", "ours", "ourselves", "you", "your", "yours", "yourself", "yourselves", "he", "him", "his", "himself", "she", "her", "hers", "herself", "it", "its", "itself", "they", "them", "their", "theirs", "themselves", "what", "which", "who", "whom", "this", "that", "these", "those", "am", "is", "are", "was", "were", "be", "been", "being", "have", "has", "had", "having", "do", "does", "did", "doing", "would", "should", "could", "ought", "i'm", "you're", "he's", "she's", "it's", "we're", "they're", "i've", "you've", "we've", "they've", "i'd", "you'd", "he'd", "she'd", "we'd", "they'd", "i'll", "you'll", "he'll", "she'll", "we'll", "they'll", "isn't", "aren't", "wasn't", "weren't", "hasn't", "haven't", "hadn't", "doesn't", "don't", "didn't", "won't", "wouldn't", "shan't", "shouldn't", "can't", "cannot", "couldn't", "mustn't", "let's", "that's", "who's", "what's", "here's", "there's", "when's", "where's", "why's", "how's", "a", "an", "the", "and", "but", "if", "or", "because", "as", "until", "while", "of", "at", "by", "for", "with", "about", "against", "between", "into", "through", "during", "before", "after", "above", "below", "to", "from", "up", "down", "in", "out", "on", "off", "over", "under", "again", "further", "then", "once", "here", "there", "when", "where", "why", "how", "all", "any", "both", "each", "few", "more", "most", "other", "some", "such", "no", "nor", "not", "only", "own", "same", "so", "than", "too", "very")


**The nchar() function counts the number of characters in a piece of text. How many characters are in the longest email in the dataset (where longest is measured in terms of the maximum number of characters)? - 43952**

In [ ]:
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [ ]:
from nltk.tokenize import RegexpTokenizer
from nltk.stem import PorterStemmer
tokenizer = RegexpTokenizer(r'\w+')
stemming = PorterStemmer()
def pre(text):
  text = text.lower()
  token = tokenizer.tokenize(text)
  ma = []
  for i in token:
    if i not in sw:
      ma.append(stemming.stem(i)) 
  return " ".join(ma)

In [ ]:
df["corpus"] = df["text"].apply(pre)
df.head()

text  spam  \
0  Subject: naturally irresistible your corporate...     1   
1  Subject: the stock trading gunslinger  fanny i...     1   
2  Subject: unbelievable new homes made easy  im ...     1   
3  Subject: 4 color printing special  request add...     1   
4  Subject: do not have money , get software cds ...     1   

                                              corpus  
0  subject natur irresist corpor ident lt realli ...  
1  subject stock trade gunsling fanni merril muzo...  
2  subject unbeliev new home made easi im want sh...  
3  subject 4 color print special request addit in...  
4  subject money get softwar cd softwar compat ai...

In [ ]:
df.shape

(5728, 3)

In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

docs = df["corpus"].tolist()
vec = CountVectorizer()
X = vec.fit_transform(docs)
dtm = pd.DataFrame(X.toarray(), columns=vec.get_feature_names())
dtm.head()


/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  0000  000000  00000000  0000000000  000000000003619  \
0   0    0     0       0         0           0                0   
1   0    0     0       0         0           0                0   
2   0    0     0       0         0           0                0   
3   0    0     0       0         0           0                0   
4   0    0     0       0         0           0                0   

   000000000003991  000000000003997  000000000005168  ...  zwzm  zxghlajf  \
0                0                0                0  ...     0         0   
1                0                0                0  ...     0         0   
2                0                0                0  ...     0         0   
3                0                0                0  ...     0         0   
4                0                0                0  ...     0         0   

   zyban  zyc  zygoma  zymg  zzmacmac  zzn  zzncacst  zzzz  
0      0    0       0     0         0    0         0     0  
1      0    0       0     0         0    0         0     0  
2      0    0       0     0         0    0         0     0  
3      0    0       0     0         0    0         0     0  
4      0    0       0     0         0    0         0     0  

[5 rows x 29236 columns]

In [ ]:
vec = CountVectorizer(min_df=0.05)
X = vec.fit_transform(docs)
spdtm = pd.DataFrame(X.toarray(), columns=vec.get_feature_names())
spdtm.head()

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  01  02  03  04  05  06  07  08  ...  week  well  will  wish  \
0   0    0   0   0   0   0   0   0   0   0  ...     0     0     6     0   
1   0    0   0   0   0   0   0   0   0   0  ...     0     0     0     0   
2   0    0   0   0   0   0   0   0   0   0  ...     0     0     0     0   
3   0    0   0   0   0   0   0   0   0   0  ...     0     0     0     0   
4   0    0   0   0   0   0   0   0   0   0  ...     0     0     0     0   

   within  without  work  write  www  year  
0       1        1     0      0    0     0  
1       0        0     0      0    0     0  
2       0        0     0      0    0     0  
3       0        0     0      0    0     0  
4       0        0     0      0    0     0  

[5 rows x 372 columns]

In [ ]:
from sklearn.model_selection import train_test_split

train, test, y_train, y_test = train_test_split(spdtm, df["spam"], test_size=0.2)



In [ ]:
from sklearn import tree # for decision tree models
spamCart = tree.DecisionTreeClassifier()
spamCart.fit(train, y_train)

DecisionTreeClassifier()

In [ ]:
print(f"test score: {spamCart.score(test, y_test)}")
print(f"training score: {spamCart.score(train, y_train)}")


test score: 0.9502617801047121
training score: 0.9986905281536447


In [ ]:
dtm.shape


(5728, 29236)

In [ ]:
from sklearn.ensemble import RandomForestClassifier
spamRF = RandomForestClassifier()
spamRF.fit(train, y_train)

RandomForestClassifier()

In [ ]:
print(f"test score: {spamRF.score(test, y_test)}")
print(f"training score: {spamRF.score(train, y_train)}")

test score: 0.9842931937172775
training score: 0.9986905281536447


In [ ]:
from sklearn import metrics
from sklearn.metrics import roc_auc_score
predCart = spamCart.predict(train)
predRF = spamRF.predict(train)
print(f"Cart AUC: {roc_auc_score(y_train, predCart)}")
print(f"RF AUC: {roc_auc_score(y_train, predRF)}")


Cart AUC: 0.9973309608540926
RF AUC: 0.9991324465008676


In [ ]:
predCart = spamCart.predict(test)
predRF = spamRF.predict(test)
print(f"Cart AUC: {roc_auc_score(y_test, predCart)}")
print(f"RF AUC: {roc_auc_score(y_test, predRF)}")

Cart AUC: 0.9220629929846245
RF AUC: 0.9780633201264948


# Conclusion
* A spam classifier can potentially benefit from including the frequency of the word that appears in every email. (Document term matrix)
* Random Forest model outperformed CART model in terms of Accuracy and AUC.
* Random forests outperform Cart because they consist of multiple single trees each based on a random sample of the training data. They are typically more accurate than single decision trees.
* While a single decision tree like CART is often pruned, a random forest tree is fully grown and unpruned.
* Each random forest tree is learned on a random sample, and at each node, a random set of features are considered for splitting. Both mechanisms create diversity among the trees which in turn give sus more accurate results.